Makefile

```diff
@@ -132,7 +132,7 @@ assets-clean:
 .PHONY: test
 # Run all unit tests
 test: spec assets
-	$(PYTHON) -m pytest --continue-on-collection-errors $(TESTDIR)
+	HOME=$(CURDIR) $(PYTHON) -m pytest --continue-on-collection-errors $(TESTDIR)
```

Comment:
```suggestion
	HOME=$(CURDIR) $(PYTHON) -m pytest --continue-on-collection-errors $(TESTDIR)
	HOME=$(CURDIR)/ocrd_utils/ocrd_utils $(PYTHON) -m pytest --continue-on-collection-errors $(TESTDIR) -k TestLogging
```

Reply:
This results on my local machine in

```
HOME=/home/m3ssman/Projects/github/ulb-sachsen-anhalt-core/ocrd_utils/ocrd_utils python -m pytest --continue-on-collection-errors tests -k TestLogging
=========================================================================================== test session starts ============================================================================================
platform linux -- Python 3.6.9, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/m3ssman/Projects/github/ulb-sachsen-anhalt-core
collected 242 items / 239 deselected / 3 selected                                                                                                                                                          

tests/test_logging.py .F.                                                                                                                                                                            [100%]

================================================================================================= FAILURES =================================================================================================
___________________________________________________________________________________ TestLogging.test_setOverrideLogLevel ___________________________________________________________________________________

self = <tests.test_logging.TestLogging testMethod=test_setOverrideLogLevel>

    def test_setOverrideLogLevel(self):
        rootLogger = logging.getLogger('')
        somelogger = getLogger('foo.bar')
        somelogger.setLevel(getLevelName('ERROR'))
>       self.assertEqual(rootLogger.getEffectiveLevel(), logging.INFO)
E       AssertionError: 10 != 20

tests/test_logging.py:20: AssertionError
========================================================================================= short test summary info ==========================================================================================
FAILED tests/test_logging.py::TestLogging::test_setOverrideLogLevel - AssertionError: 10 != 20
=============================================================================== 1 failed, 2 passed, 239 deselected in 0.91s ================================================================================
Makefile:135: recipe for target 'test' failed
make: *** [test] Error 1

```

Reply:
Something different ... I needed to add some python libs to `requirements_test.txt`, shall I propagate this?

Reply:
Feel free to add to `requirements_test.txt` but keep in mind that these are installed for every run of CI. Which libs do you want to add?

Reply:
Before I could run `make test` locally without lib-errors, I needed to add `atomicwrites`, `deprecated` and `opencv-python`. The last library is quite large, though.

Reply:
@kba Please nevermind my last remarks. Additionally, I also need to do install `ocrd_utils` because of the pushd_popd stuff. After all, I don't think it is desirable and I really like to come to an end with this ;)

Reply:
These libs are already in `requirements.txt` (as `Deprecated`, `atomicwrites` and `opencv-headless-python`). Are you sure, you've run `pip install -e .`?

Reply:
I get the same assertion error as @m3ssman with the proposed change to Makefile BTW.

Reply:
@bertsky 
The second test locally only works if I add an explicit `initLogging`, say in a `setUp` method. Otherwise, it alerts like posted above and fails.

Reply:
> E       AssertionError: 10 != 20

That's interesting. It means that the effective level of the root logger is not `INFO` (20) but `DEBUG` (10; `TRACE`) here.

But we never set it to that level anywhere! (Except in `test_decorators.TestDecorators.test_loglevel_override`.)

Next observation: this happens regardless of what `HOME` points to. (Also, in my first suggestion, I mistakenly assumed the template was under `ocrd_utils/ocrd_utils/ocrd_logging.conf`, whereas in fact it resides one directory level upwards. So this had no effect anyway.)

Obviously, just picking one testcase/class via `-k` invalidates our setup.

Reply:
> Obviously, just picking one testcase/class via `-k` already invalidates our setup.
> But we never set it to that level anywhere! (Except in `test_decorators.TestDecorators.test_loglevel_override`.)

We're lucky to have stumbled over this at last.

Wow. The problem is that we have a `setOverrideLogLevel('DEBUG')` in `test_workspace` – on the **module level** (i.e. without even starting any test). This is also in effect when filtering test classes/cases via `-k`, because obviously pytest needs to first discover what's there before it can filter.

This was introduced by 53ca390a – and looks like an accident.

But that reminds us that we also need a much more strict test case isolation.

Reply:
> But that reminds us that we also need a much more strict test case isolation.

So how do we undo the effects of `setOverrideLogLevel` – ever? We cannot access `ocrd_utils.logging._overrideLogLevel`, because it is not exported. But `setOverrideLogLevel` does not allow to set it back to `None`.

Reply:
> This was introduced by 53ca390 – and looks like an accident.

Indeed it was, should be reverted.

> So how do we undo the effects of setOverrideLogLevel

Accept `None` as argument and if so, set root logger to `NOTSET`?

Reply:
> So how do we undo the effects of `setOverrideLogLevel` – ever (or of `initLogging`'s `basicConfig`, for that matter)? We cannot access `ocrd_utils.logging._overrideLogLevel`, because it is not exported. But `setOverrideLogLevel` does not allow to set it back to `None`.

I suggest the following:
- `initLogging` should attempt to undo both any previous `basicConfig`, by [removing all root logger handlers](https://stackoverflow.com/a/12158233), and then set `_overrideLogLevel = None`. Now we can reliably re-init by simply calling it again.
- to get test case isolation, we should create a `setUp` implementation (or `setUpClass`) including `initLogging` for all test cases that need a certain logging state. Also, test cases that manipulate the logging state for their own purposes should probably have a `tearDown` implementation (or `tearDownClass`) likewise.

Reply:
> > This was introduced by [53ca390](https://github.com/OCR-D/core/commit/53ca390a301cdfae0e1b3a71a6426fac6dbbd3ce) – and looks like an accident.
> 
> Indeed it was, should be reverted.

The whole commit, including `test_download_to_directory_from_workspace_download_file`? Or just the `setOverrideLogLevel` (and probably the `system('find %s' % ws_dir)`)?

Reply:
Again, thanks for digging and proposing the fixes :) `initLogging` resetting all overrides makes sense to me.

Have a look at https://github.com/kba/ocrd-core/commit/bd232c0dfc7363538957e5db67e0221aef89ac3e That should revert the debug statements, changes the TestCase base class to do `initLogging()` per default on `tearDown` and change initLogging() to remove all handlers for root loggers and reset `_overrideLogLevel` to `None`.

@m3ssman Feel free to use that commit, I didn't want to push to your master branch in case you're working on this.